gem/optimise.py

```diff
@@ -1029,3 +1029,77 @@ def propogate_failure(expressions):
         if isinstance(n, Failure):
             return True
     return False
+
+
+class Kernel_Optimiser(object):
```

Comment:
Please use names that are in conformance with the standard Python conventions. They are not that hard to learn:

- Camel case (no underscode) starting with capital letter for class names. The only exception here are decorators.
- Lower case (possibly separated by underscores) basically in all other cases: function names, variables, module names, etc.